### PR TITLE
Object builder API updates

### LINKED ISSUE #988: FabricBlockSettings.of overload with Function<BlockState, MaterialColor>
Vanilla has this in `AbstractBlock.Settings` (which I can use in my case), but it'd be useful for mods as well.

### PR DESCRIPTION
**Pull request by apple502j**
- Add a way to specify required features for entity types, like for blocks and items.
- Resolves #988
- Misc fixes.
- GameTest.

### CODE REVIEW

File: fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/FabricEntityTypeBuilder.java [hunk 1]
```diff
@@ -250,20 +254,26 @@ public FabricEntityTypeBuilder<T> specificSpawnBlocks(Block... blocks) {
 		return this;
 	}
 
+	/**
+	 * Sets the features this entity requires. If the feature is not enabled,
```
Comment (Juuxel) [inline fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/FabricEntityTypeBuilder.java]:
```suggestion
	 * Sets the features this entity requires. If a feature is not enabled,
```
since we're not talking about just one feature here

File: fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/MinecartComparatorLogicRegistry.java [hunk 2]
```diff
@@ -16,23 +16,24 @@
...
 import net.minecraft.entity.vehicle.AbstractMinecartEntity;
+import net.minecraft.registry.Registries;
 
 /**
  * A registry for {@linkplain MinecartComparatorLogic custom minecart comparator logic}.
  */
 public final class MinecartComparatorLogicRegistry {
 	private static final Logger LOGGER = LoggerFactory.getLogger(MinecartComparatorLogicRegistry.class);
-	private static final Map<EntityType<?>, MinecartComparatorLogic<?>> logics = new HashMap<>();
+	private static final Map<EntityType<?>, MinecartComparatorLogic<?>> logics = new IdentityHashMap<>();
```
Comment (Juuxel) [inline fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/MinecartComparatorLogicRegistry.java]: Since you're changing it anyway, should be all-caps since "All names should follow the [Yarn naming standards](https://github.com/FabricMC/yarn/blob/HEAD/CONVENTIONS.md)." (CONTRIBUTING.md)

File: fabric-object-builder-api-v1/src/testmod/java/net/fabricmc/fabric/test/object/builder/VillagerTypeTest1.java [hunk 3]
```diff
@@ -49,12 +49,6 @@ public void onInitialize() {
 		});
 
 		CommandRegistrationCallback.EVENT.register((dispatcher, registryAccess, environment) -> {
-			dispatcher.register(literal("fabric_refreshtrades").executes(context -> {
-				TradeOfferHelper.refreshOffers();
-				context.getSource().sendFeedback(Text.literal("Refreshed trades"), false);
-				return 1;
-			}));
```
Comment (Juuxel) [inline fabric-object-builder-api-v1/src/testmod/java/net/fabricmc/fabric/test/object/builder/VillagerTypeTest1.java:56]: Hmm, why was this removed

Comment (apple502j) [reply fabric-object-builder-api-v1/src/testmod/java/net/fabricmc/fabric/test/object/builder/VillagerTypeTest1.java:56]: The method does nothing. (Should this be `forRemoval=true`'d)?

**Review by Juuxel — APPROVED** [review]

**Comment by Technici4n** [comment]
Can you also include a fix for #988?

File: fabric-object-builder-api-v1/src/testmod/java/net/fabricmc/fabric/test/object/builder/VillagerTypeTest1.java (hunk 3 above)
Comment (Technici4n) [reply fabric-object-builder-api-v1/src/testmod/java/net/fabricmc/fabric/test/object/builder/VillagerTypeTest1.java:56]: Yeah it was deprecated in 1.16 before `forRemoval=true` was a thing, see #1430. This is one of the instances where the method is so broken we should deprecate it for removal.

File: fabric-object-builder-api-v1/src/testmod/resources/fabric.mod.json [hunk 4]
```diff
@@ -26,6 +26,9 @@
       "net.fabricmc.fabric.test.object.builder.CriterionRegistryTest::init",
       "net.fabricmc.fabric.test.object.builder.VillagerTypeTest1",
       "net.fabricmc.fabric.test.object.builder.VillagerTypeTest2"
+    ],
+    "fabric-gametest": [
+      "net.fabricmc.fabric.test.object.builder.BlockEntityTypeBuilderTest"
```
Comment (Technici4n) [inline fabric-object-builder-api-v1/src/testmod/resources/fabric.mod.json]: No idea if putting the game tests in the same file as the mod initializer is good practice (doesn't seem so).

File: fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/FabricEntityTypeBuilder.java [hunk 5]
```diff
@@ -55,6 +57,8 @@
 	private EntityDimensions dimensions = EntityDimensions.changing(-1.0f, -1.0f);
 	private ImmutableSet<Block> specificSpawnBlocks = ImmutableSet.of();
 
+	private FeatureSet requiredFeatures = FeatureFlags.VANILLA_FEATURES;
```
Comment (Technici4n) [inline fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/FabricEntityTypeBuilder.java:60]: Shouldn't this be `DEFAULT_ENABLED_FEATURES` instead?

File: fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/MinecartComparatorLogicRegistry.java [hunk 6]
```diff
@@ -57,9 +58,13 @@ public static MinecartComparatorLogic<AbstractMinecartEntity> getCustomComparato
 	 * @param <T>   the handled minecart type
 	 * @param type  the minecart entity type
 	 * @param logic the logic to register
+	 * @throws NullPointerException if the type or the logic is {@code null}
```
Comment (Technici4n) [inline fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/MinecartComparatorLogicRegistry.java]: Bonk

File: fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/FabricEntityTypeBuilder.java (hunk 5 above)
Comment (apple502j) [reply fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/FabricEntityTypeBuilder.java:60]: Hmm, either works.

File: fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/MinecartComparatorLogicRegistry.java (hunk 6 above)
Comment (apple502j) [reply fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/entity/MinecartComparatorLogicRegistry.java]:
```suggestion
```

**Review by Technici4n — APPROVED** [review]

**Review by modmuss50 — APPROVED** [review]
